Add value counts under polygons in zonal statistics

phargogh opened the pull request:
This PR adds pixel counts under each polygon and warns the user if pixel counts have been requested on a floating-point raster.

Fixes #270

phargogh committed:
Adding this feature and starting a test for it.

RE:#270

phargogh committed:
Updating Rich's email address to mine since his no longer works. RE:#270

phargogh committed:
Minor grammar fix, object access fix and improving assertions. RE:#270

phargogh committed:
Noting new feature in HISTORY. RE:#270

phargogh committed:
Merge branch 'main' of github.com:natcap/pygeoprocessing into feature/270-value-counts-under-polygons-in-zonal-statistics

phargogh committed:
Correcting HISTORY - not part of 2.3.4 release.

RE:#270

phargogh committed:
More explicitly handling possibility of no matches in a block.

RE:#270

phargogh committed:
Removing a couple LOGGER messages. RE:#720

phargogh committed:
Reworking defaultdict initializer to use a Counter.

A `Counter` object allows for succinct tallying of landuse codes, so it
really makes sense to use in this case.

RE:#270

phargogh committed:
Minor tweak to language in logging message. RE:#270

phargogh committed:
Correcting variable name. RE:#270

phargogh committed:
Minor linting for line length. RE:#270

phargogh commented:
@emlys this should be ready to review now!

emlys commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -1234,12 +1244,39 @@ def zonal_statistics(
         temp_working_dir, 'clipped_raster.tif')
 
+    def _make_sample_dict():
+        """Build a sample aggregate dict.
```
I'm not sure I'm understanding why it's a "sample" dict?

phargogh replied:
Yeah, "sample" might not be the best terminology here ... "new" is better.  I've revised this in 8aa8227 to clarify what this is used for.  "sample" was borrowed from the prior implementation where the term maybe made a little more sense based on how it was used, as a starter value.

emlys commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -1234,12 +1244,39 @@ def zonal_statistics(
+            'min': None, 'max': None, 'count': 0, 'nodata_count': 0,
+            'sum': 0.0}
+        if include_value_counts:
+            sample_aggregate_dict['value_counts'] = collections.Counter()
```
neat!

phargogh replied:
I think this is the first time I've maybe ever needed to use a counter!  Pretty exciting.

emlys commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -1194,7 +1199,12 @@ def zonal_statistics(
+                 'value_counts': {
+                     2: 5,
+                     4: 1,
+                     14: 2,
```
minor, but it would be nice if these example statistics were consistent with each other

phargogh replied:
Yes it would!  Patched in e32064c

emlys commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -1487,6 +1524,10 @@ def zonal_statistics(
         aggregate_stats[unset_fid]['nodata_count'] = numpy.count_nonzero(
             unset_fid_nodata_mask)
+        if include_value_counts:
+            aggregate_stats[unset_fid]['value_counts'].update(
```
this might be worth a comment that `aggregate_stats[unset_fid]['value_counts']` is a `collections.Counter` and so `update` adds to it rather than overwrites.

phargogh replied:
I agree, patched in 7dccc77

phargogh committed:
Renaming the defaultdict factory function and vars.

I'm trying to clarify the intent of the function and the created dict
... "sample" isn't a good descriptor for what's going on here.

RE:#270

phargogh committed:
Updating example stats output to be consistent. RE:#270

phargogh committed:
Adding a comment about how update() is different on a Counter object. RE:#270

emlys reviewed (approved):
thanks @phargogh!
